README.md
Wait for some HTTP response
===========================

Written in order to make functional tests less complicated.

Retries a HTTP request until it responds as expected or reaches
the max attempt number.

------------------------------------------------------------------

Lets say we have an endpoint `http://localhost:8080/health` that give
us a json, e.g., `{ "status": "UP" }`. This script keep checking that endpoint
until it receives the `UP` status or it reaches the max number of attempts.

# Install

`pip install wait_response`

# Module

You can use a function from the module to wait for status responses.

```python
import wait_response
from wait_response import wait_response_status

# Make 2 attempts to get status=UP, trying every 1 seconds. 
respCode = wait_response_status('http://localhost:8080/health', 2, 1, 'UP')
print(respCode) # 0 if success or else, 1

```

# Script

`wait_response url [OPTIONS]`

`url` is the required endpoint, e.g., `http://localhost:8080/health`.

## Options

* `--max-attempts default=20` The max number of attempts.
* `--sleep defaul=1` Sleep time, in seconds, between checks.
* `--status default=UP` The status to wait for. E.g., `GREEN` or `YELLOW`.

# Developing

## Tests

Use `unittest`
`python -m unittest test.test_wait_response`

## Run script


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Add a way to specify any expected response in function `wait_response_status`.

## [1.0.0-rc1] - 2017-10-30
### Added
- Function `wait_response_status` for waiting to `{"status": "SOME_VALUE"}` response.
- Script functionality.


test/test_waitresponse.py
import unittest

import responses

import waitresponse
from waitresponse import waitresponse

class TestWaitresponse(unittest.TestCase):

    def test_waitresponse_until_max_attempts(self):
        result = waitresponse('http://localhost:8080/someWrongUrl', 2, 1, 'UP')
        self.assertEqual(result, 1)

    @responses.activate
    def test_waitresponse_ok(self):
        responses.add(responses.GET, 'http://localhost:8080/health',
                  json={'status': 'UP'}, status=200)
        result = waitresponse('http://localhost:8080/health', 2, 1, 'UP')
        self.assertEqual(result, 0)

if __name__ == '__main__':
    unittest.main()


test/test_wait_response.py
#!/usr/bin/env python3

import unittest

import responses

from wait_response import wait_response_status


class TestWaitResponse(unittest.TestCase):
    def test_wait_response_status_until_max_attempts(self):
        result = wait_response_status('http://localhost:8080/someWrongUrl', 2, 1, 'UP')
        self.assertEqual(result, 1)

    @responses.activate
    def test_wait_response_status_up(self):
        responses.add(responses.GET, 'http://localhost:8080/health',
                      json={'status': 'UP'}, status=200)
        result = wait_response_status('http://localhost:8080/health', 2, 1, 'UP')
        self.assertEqual(result, 0)


if __name__ == '__main__':
    unittest.main()


waitresponse/waitresponse.py
#!/usr/bin/python

from time import sleep
from requests import get as http_get

URL, ATTEMPTS, SLEEP_SECS, STATUS_UP = 'http://localhost:8080/health', 20, 1, 'UP'

def health_match_status(url, status):
    r = http_get(url)
    if r.status_code is not 200:
        return False
    return r.json()['status'] is not status


def waitresponse(url=URL, max_attempts=ATTEMPTS, sleep_secs=SLEEP_SECS, status=STATUS_UP):
    # Indicates if the status from the response is the requested
    matched = False

    # Runs once + retry times
    for i in range(max_attempts):
        try:
            matched = health_match_status(url, status)
            if matched:
                break
        except Exception as err:
            print("Attempt {0}/{1} error: {2}".format(i + 1, max_attempts, err))
        # Does not need to sleep because it will not try again
        if i is not max_attempts - 1:
            sleep(sleep_secs)

    if matched is True:
        return 0
    else:
        return 1


wait_response/wait_response.py
#!/usr/bin/env python
# coding: utf-8

from argparse import ArgumentParser
from time import sleep
from requests import get as http_get

URL, ATTEMPTS, SLEEP_SECS, STATUS_UP = 'http://localhost:8080/health', 20, 1, 'UP'


def health_match_status(url, status):
    r = http_get(url)
    if r.status_code is not 200:
        return False
    return r.json()['status'] is not status


def wait_response_status(url=URL, max_attempts=ATTEMPTS, sleep_secs=SLEEP_SECS, status=STATUS_UP):
    # Indicates if the status from the response is the requested
    matched = False

    # Runs once + retry times
    for i in range(max_attempts):
        try:
            matched = health_match_status(url, status)
            if matched:
                break
        except Exception as err:
            print("Attempt {0}/{1} error: {2}".format(i + 1, max_attempts, err))
        # Does not need to sleep because it will not try again
        if i is not max_attempts - 1:
            sleep(sleep_secs)

    if matched is True:
        return 0
    else:
        return 1


def main():
    parser = ArgumentParser()

    parser.add_argument('url',
                        help='The status endpoint. E.g., http://localhost:8080/health')
    parser.add_argument('--max-attempts',
                        help="The max number of attempts. (default: {0})".format(ATTEMPTS),
                        type=int,
                        default=ATTEMPTS)
    parser.add_argument('--sleep',
                        help='Sleep time, in seconds, between checks. (default: 1)',
                        default=SLEEP_SECS,
                        type=int)
    parser.add_argument('--status',
                        default=STATUS_UP,
                        help='The status to wait for. E.g., `GREEN` or `YELLOW`. (default: UP)')
    args, unknown_args = parser.parse_known_args()

    exit(wait_response_status(args.url, args.max_attempts, args.sleep, args.status))


if __name__ == "__main__":
    main()
